crates/turborepo-ui/src/tui/preferences.rs

```diff
@@ -1,132 +1,73 @@
-        repo_root: &AbsoluteSystemPathBuf,
-        field: PreferenceFields,
-        new_value: Value,
-    ) -> Result<(), Box<dyn std::error::Error>> {
-        // TODO: Clean these up, should be taken from constants
-        let preferences_dir = repo_root.join_components(&[".turbo", "preferences"]);
-        let preferences_file = repo_root.join_components(&[".turbo", "preferences", "tui.json"]);
-
-        fs::create_dir_all(preferences_dir.as_std_path())?;
-
-        let mut json: Value = if preferences_file.exists() {
-            let json_string = fs::read_to_string(&preferences_file)?;
-            serde_json::from_str(&json_string)?
-        } else {
-            json!({})
-        };
-
-        // TODO: Is this really how to do this? No way, right?
-        match field {
-            PreferenceFields::IsTaskListVisible => {
-                json["is_task_list_visible"] = new_value;
-            }
-            PreferenceFields::ActiveTask => {
-                json["active_task"] = new_value;
-            }
-            PreferenceFields::PinnedTaskSelection => {
-                json["is_pinned_task_selection"] = new_value;
-            }
-        }
 
-        let updated_json_string = serde_json::to_string_pretty(&json)?;
-
-        let mut file = fs::File::create(&preferences_file)?;
-        file.write_all(updated_json_string.as_bytes())?;
-
-        Ok(())
+    pub fn is_task_list_visible(&self) -> bool {
+        self.config.is_task_list_visible.unwrap_or(true)
     }
 
-    pub fn read_pinned_task_state(repo_root: &AbsoluteSystemPathBuf) -> bool {
-        let preferences_file = repo_root.join_components(TUI_PREFERENCES_PATH_COMPONENTS);
-
-        read_json(&preferences_file)
-            .is_pinned_task_selection
-            .unwrap_or(false)
+    pub fn active_task(&self) -> Option<&str> {
+        let active_task = self.config.active_task.as_deref()?;
+        Some(active_task)
     }
 
-    pub fn read_task_list_visibility(repo_root: &AbsoluteSystemPathBuf) -> bool {
-        let preferences_file = repo_root.join_components(TUI_PREFERENCES_PATH_COMPONENTS);
-
-        read_json(&preferences_file)
-            .is_task_list_visible
-            .unwrap_or(true)
+    pub fn set_is_task_list_visible(&mut self, value: Option<bool>) -> Result<(), Error> {
+        self.config.is_task_list_visible = value;
+        self.flush_to_disk()
     }
 
-    pub fn get_selected_task_index(
-        repo_root: &AbsoluteSystemPathBuf,
-        tasks_by_status: &TasksByStatus,
-    ) -> usize {
-        let preferences_file = repo_root.join_components(TUI_PREFERENCES_PATH_COMPONENTS);
+    pub fn set_active_task(&mut self, value: Option<String>) -> Result<(), Error> {
+        self.config.active_task = value;
+        self.flush_to_disk()
+    }
 
-        let selected_task_name = read_json(&preferences_file)
-            .active_task
-            .unwrap_or("".to_string());
+    fn flush_to_disk(&self) -> Result<(), Error> {
```

Comment:
Right now we're flushing to disk everytime one of these configurations change. Could we get away with only doing this when the TUI is shutting down? (If not we need to spin up a separate thread to do this so we're not performing FS operations on the render thread)

Reply:
This was something I'd thought about but didn't realize `cleanup()` was there on `App`. Yes, very doable.